src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -26,73 +26,61 @@ private void FinalizeDispose()
-                                    Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed;
-                }
+                return;
+            }
+            else
+            {
+                _cancellation = new CancellationTokenSource();
+            }
 
-                CreateStreamAndStartWatcher();
+            Interop.EventStream.FSEventStreamEventFlags flags = 0;
+
+            // Always re-create the filter flags when start is called since they could have changed
+            if ((_notifyFilters & (NotifyFilters.Attributes | NotifyFilters.CreationTime | NotifyFilters.LastAccess | NotifyFilters.LastWrite | NotifyFilters.Size)) != 0)
+            {
+                flags = Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemInodeMetaMod  | 
+                               Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemFinderInfoMod |
+                               Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemModified      |
+                               Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemChangeOwner;
             }
+            if ((_notifyFilters & NotifyFilters.Security) != 0)
+            {
+                flags |= Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemChangeOwner | Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemXattrMod;
+            }
+            if ((_notifyFilters & NotifyFilters.DirectoryName) != 0)
+            {
+                flags |= Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsDir |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsSymlink |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed;
+            }
+            if ((_notifyFilters & NotifyFilters.FileName) != 0)
+            {
+                flags |= Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsSymlink |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved |
+                                Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed;
+            }
+
+            RunningInstance instance = new RunningInstance(this, _directory, _includeSubdirectories, flags, _cancellation.Token);
+            instance.Start();
+            _enabled = true;
```

Comment:
_enabled needs to be set to true before instance.Start is called.  Otherwise, I believe there are some theoretical race conditions that could result in problems.  Then since Start can throw an exception (it doesn't in the Linux implementation), _enabled will need to be set back to false in a catch block.  Similarly, _cancellation will need to be set back to null if either the RunningInstance ctor or Start throws... this is why the Linux implementation doesn't set _cancellation to non-null until the same time it sets _enabled to true.